rfd/0031-dynamic-app-db-registration.md

```diff
@@ -0,0 +1,355 @@
+
+```bash
+$ tctl create grafana-app.yaml
+$ tctl create redshift-db.yaml
+$ tctl get app_servers
+$ tctl get db_servers
+$ tctl get app_servers/grafana
+$ tctl rm db_servers/redshift
+```
+
+**Refactoring & implementation notes**
+
+- Apps are currently a part of the [Server](https://github.com/gravitational/teleport/blob/v7.0.0-dev.9/api/types/types.proto#L232)
+  object. As a part of this work, refactor it into a separate `AppServer` object
+  similar to [DatabaseServer](https://github.com/gravitational/teleport/blob/v7.0.0-dev.9/api/types/databaseserver.go).
+- Rename apps `labels` and `commands` to `static_labels` and `dynamic_labels`
+  respectively, internally and in file configuration, in backwards compatible
+  fashion.
+- App/database resources currently have kinds `app_server` and `db_server`
+  respectively. As a part of this cleanup, we can alias them to `app` and `db`
+  which would be cleaner (but require some extra work).
+
+## Configuring app/database service
+
+In addition to apps/databases from static yaml configuration, app/database
+services can be configured to watch for the above resources with specific
+labels. If no selectors are specified, no resources are being watched.
+
+```yaml
+app_service:
+  enabled: "yes"
+  # Watch for apps with app=ops&env=dev or env=test labels.
+  selectors:
+  - matchLabels:
+      app: ops
+      env: dev
+  - matchLabels:
+      env: test
+  # Existing static apps configuration, may be omitted.
+  apps:
+    ...
+```
+
+```yaml
+db_service:
+  enabled: "yes"
+  # Watch for any database with env label set.
+  selector:
+  - matchLabels:
+      env: "*"
+  # Existing static databases configuration, may be omitted.
+  dbs:
+    ...
+```
+
+The `matchLabels` approach is inspired by Kubernetes selectors. It is the only
+type of selector supported initially, but gives us a way to introduce other
+types in future if necessary (e.g. K8s has `matchExpression`).
+
+To make a service watch for any app/database, an explicit wildcard selector
+must be specified:
+
+```yaml
+selector:
+- matchLabels:
+    "*": "*"
+```
+
+**Static vs dynamic configuration**
+
+There must be a way to distinguish between apps and databases from the static
+yaml configuration vs ones registered as dynamic resources. This is needed,
+for example, to prevent users from deleting those registered statically.
+
+We will use the same approach as in [RFD16 Dynamic Configuration](./0016-dynamic-configuration.md)
+and use `teleport.dev/origin` label to denote whether the app/database resource
+comes from static configuration or created dynamically. The rules are:
+
+- Users can't delete apps/databases with `teleport.dev/origin: config-file`
+  label.
+- Users can't set `teleport.dev/origin` label in their resources - it will be
+  auto-set to `dynamic` when they're created.
+- Users can't create apps/databases with the same name as one of existing
+  static apps/databases (to avoid various conflict resolution issues).
```

Comment:
what happens in the opposite case - when a dynamic app/db exists and a new service is starting with static config using the same name?